eclipse/capella-basic-vp
# Move mass computation to pre-commit listener

### Problem
Mass computations for each element of the diagram are called from a conditional style in the odesign file. This means that every time the view is refreshed, the mass of every element of the diagram is re-calculated, leading to performance issues on bigger diagrams.

### Solution
I provide a modification relying on a pre-commit listener on the resource manager of the session to trigger mass computations. This listener is only triggered when adding a mass to the model, editing the mass of an element, moving an element with a mass, or removing an element with a mass.
Furthermore, the computation is only called on the parents/child of the triggering element, and not on every element of the diagram like it previously was.

### Validation
I used the following diagram to compare both approaches :

![DiagramBefore](https://user-images.githubusercontent.com/61457677/76500251-81684400-6440-11ea-8a3e-9393a3a16ce3.PNG)

I then followed this scenario on this diagram : 

1. Change name of PC 1 to PC 12
2. Change PC1.2 mass to 50
3. Change PC 12 mass to 10
4. Create PC 1.5.1.1, child of PC 1.5.1
5. Create PC 1.3.1, child of PC 1.3
6. Add mass to PC 1.5.1.1
7. Add mass to PC 1.3.1
8. Change PC 1.5.1.1 mass to 10
9. Change PC 1.3.1 mass to 10

To obtain this one : 
![DiagramAfterManip2](https://user-images.githubusercontent.com/61457677/76500656-37cc2900-6441-11ea-8efc-5f495116a9a3.PNG)

I used [Yourkit](https://www.yourkit.com/) to compare how many times both approaches were calling the computeMass method : 
1. With the computation being triggered directly from the odesign, the method was called 276 times : 
![image](https://user-images.githubusercontent.com/61457677/76501315-554dc280-6442-11ea-880d-2feab640f03d.png)
2. With the computation being triggered by the listener approach, the method was called 24 times : 
![image](https://user-images.githubusercontent.com/61457677/76501623-e6249e00-6442-11ea-9c4d-3692dd3c980c.png)

I also added a new build profile that lauches a JUnit test suite. This test suite contains 9 test cases to check that the mass of the elements of the diagram are correctly computed during the following scenarios : 
1. Opening a new session
2. Adding a new mass to the model
3. Removing a mass from the model
4. Editing a mass from the model
